presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSubsetNodeSetSupplier.java

```diff
@@ -0,0 +1,340 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.execution.scheduler.NodeSelectionHashStrategy.CONSISTENT_HASHING;
+import static com.facebook.presto.metadata.InternalNode.NodeStatus.ALIVE;
+import static com.facebook.presto.spi.NodeState.ACTIVE;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.lang.Math.ceil;
+import static java.lang.String.format;
+
+/**
+ * This WIP class is responsible for assigning a subset of the cluster nodes
+ * to a requesting query.
+ * The size of the subset is currently static, but it can by defined by configuration
+ * or can be driven by fairness algorithm depending on query metadata.
+ */
+public class FixedSubsetNodeSetSupplier
+        implements NodeSetSupplier, Closeable
+{
+    private static final Logger log = Logger.get(FixedSubsetNodeSetSupplier.class);
+    private final PriorityQueue<NodeSetAcquireRequest> pendingRequests;
+    private final InternalNodeManager nodeManager;
+    private final NetworkLocationCache networkLocationCache;
+    // TODO: Use Set<InternalNode>
+    private final Map<QueryId, List<InternalNode>> queryNodesMap;
+
+    private final Map<QueryId, Supplier<QueryState>> queryStateMap;
+
+    private final ScheduledFuture<?> scheduledFuture;
+
+    @Inject
+    public FixedSubsetNodeSetSupplier(
+            InternalNodeManager nodeManager,
+            NetworkTopology networkTopology,
+            @ForNodeScheduler ScheduledExecutorService scheduledExecutorService)
+    {
+        this.nodeManager = nodeManager;
+        this.networkLocationCache = new NetworkLocationCache(networkTopology);
+        this.queryNodesMap = new ConcurrentHashMap<>();
+        this.queryStateMap = new ConcurrentHashMap<>();
+        this.pendingRequests = new PriorityQueue<>();
+        // start loop to check pending node requests every second
+        scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(this::fulfillPendingRequests, 0, 100, TimeUnit.MILLISECONDS);
+    }
+
+    private void fulfillPendingRequests()
+    {
+        if (pendingRequests.size() == 0) {
+            return;
+        }
+
+        // In case we have not released nodes on finished queries.
+        // This is to prevent leaking nodes over time as falsely occupied
+        for (Map.Entry<QueryId, List<InternalNode>> entry : queryNodesMap.entrySet()) {
+            if (queryStateMap.containsKey(entry.getKey()) && queryStateMap.get(entry.getKey()).get().isDone()) {
+                releaseNodes(entry.getKey(), entry.getValue().size());
+            }
+        }
+
+        List<InternalNode> freeNodes = computeFreeNodesInCluster();
```

Comment:
Should this min workers ? Like skip fullfilling if the num free nodes is less than what waitForMinWorkers originally did -- to wait for clusters to "turn up"

Reply:
Yes. Responded to this in cluster turn up use-case